### PR TITLE
PCRE2

### PR DESCRIPTION
**Pull request by evgenyz**
This PR introduces support for PCRE2 library for regular expression handling. It does it in a backward-compatible way, keeping the possibility of linking with now-obsolete PCRE1 library.

No changes in the code are required to switch between libraries, the `oscap_pcre_*` group of functions would take care of all the differences between compilation and matching API.

By default the build system will try to link with the old PCRE1 library. To force the new PCRE2 one must pass `-DWITH_PCRE2=True` argument during CMake configuration step.

This PR also moves `oscap_get_substrings` function from `utils` into `oscap_pcre` module, because of the semantics and implicit dependencies of this function.

It also get rid of code that handles `PCRE_ERROR_BADPARTIAL` match error, as this error is not actual for any usable version of the library.

### CODE REVIEW

File: src/OVAL/probes/oval_fts.c [hunk 1]
```diff
@@ -577,35 +446,15 @@ static int process_pattern_match(const char *path, pcre **regex_out)
 		}
 		*/
 		break;
-	case PCRE_ERROR_BADPARTIAL:
```
Comment (evgenyz) [inline src/OVAL/probes/oval_fts.c:580]: We don't need that case anymore. Any recent version of the library does not have limits on partial matches. This is just a dead code block now.

File: tests/probes/textfilecontent54/test_recursion_limit.sh [hunk 2]
```diff
@@ -16,7 +16,7 @@ stderr=$(mktemp)
 
 $OSCAP oval eval --results $result $input > $stdout 2> $stderr
 
-grep -q "Function pcre_exec() failed to match a regular expression with return code -21" $stderr
+grep -q "Function oscap_pcre_exec() failed to match a regular expression with return code -21" $stderr
```
[the remark this answers is not in the document]
Comment (evgenyz) [reply tests/probes/textfilecontent54/test_recursion_limit.sh:19]: No, but `oscap_pcre_exec` will.

File: openscap.spec [hunk 3]
```diff
@@ -18,7 +18,7 @@ BuildRequires:  apt-devel
 %endif
 BuildRequires:  rpm-devel
 BuildRequires:  libgcrypt-devel
-BuildRequires:  pcre-devel
+BuildRequires:  pcre2-devel
```
[the remark this answers is not in the document]
Comment (evgenyz) [reply openscap.spec:22]: I confined it to Fedoras + upstream Ubuntu.

File: openscap.spec [hunk 4]
```diff
@@ -136,6 +136,7 @@ Tool for scanning Atomic containers.
 # gconf is a legacy system not used any more, and it blocks testing of oscap-anaconda-addon
 # as gconf is no longer part of the installation medium
 %cmake \
+    -DWITH_PCRE2=ON \
```
[the remark this answers is not in the document]
Comment (evgenyz) [reply openscap.spec:144]: Ditto, Fedora+Ubuntu only.

File: docs/developer/developer.adoc [hunk 5]
```diff
@@ -319,7 +319,7 @@ behaviour.
 * *OSCAP_FULL_VALIDATION=1* - validate all exported documents (slower)
 * *SEXP_VALIDATE_DISABLE=1* - do not validate SEXP expressions (faster)
 * *OSCAP_PCRE_EXEC_RECURSION_LIMIT* - override default recursion limit
-  for match in pcre_exec call in textfilecontent(54) probes.
+  for match in pcre_exec/pcre2_match calls in textfilecontent(54) probes.
```
[the remark this answers is not in the document]
Comment (evgenyz) [reply docs/developer/developer.adoc:322]: By default the build system will try to find and link with PCRE1 library (minor version, we should not break the behavior). Changing the requirements in the documentation could bring a lot of confusion.

File: tests/API/probes/oval_fts_list.c [hunk 6]
```diff
@@ -172,6 +176,8 @@ int main(int argc, char *argv[])
 
 	int ret = 0;
 
+	oscap_set_verbose(NULL, NULL);
```
[the remark this answers is not in the document]
Comment (evgenyz) [reply tests/API/probes/oval_fts_list.c:179]: The first `NULL` in `oscap_set_verbose` sets the default log level (WARNING) and the second sets the default output (STDERR). Without this change the test binary won't display any logs. Very inconvenient.

File: src/common/oscap_pcre.c [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+
+oscap_pcre_t* oscap_pcre_compile(const char *pattern, oscap_pcre_options_t options,
+                                 char **errptr, int *erroffset)
+{
+	oscap_pcre_t *res = malloc(sizeof(oscap_pcre_t));
+#ifdef HAVE_PCRE2
+	int errno;
+	PCRE2_SIZE erroffset2;
+	res->re_ctx = NULL;
+	dD("pcre2_compile_8: patt=%s", pattern);
```
[the remark this answers is not in the document]
Comment (evgenyz) [reply src/common/oscap_pcre.c:145]: I have a feeling that we will need these regex-related logs for debugging problems very soon. It's debug level anyways.

File: src/common/oscap_pcre.c [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+	case PCRE_ERROR_RECURSIONLIMIT:
+		return OSCAP_PCRE_ERR_RECURSIONLIMIT;
+#endif
+	default:
+		dW("Unknown PCRE error code: %d", res);
+		return OSCAP_PCRE_ERR_UNKNOWN;
+	}
+}
+
+oscap_pcre_t* oscap_pcre_compile(const char *pattern, oscap_pcre_options_t options,
```
[the remark this answers is not in the document]
Comment (evgenyz) [reply src/common/oscap_pcre.c:137]: Fixed.

**Review by jan-cerny — CHANGES_REQUESTED** [review]
Overall, the changes look OK to me. Thanks for your patch!

I have reviewed the build and test results and I have seen that we build and test with PCRE2 in GH CI.

Regarding the Developer Guide, I understand your point, but I think that you should mention it there that there is an option to build the project with PCRE2. You can also make the Developer Guide ready for the situation that PCRE will be sunset in future Fedora.

Lastly, please fix the Windows build, we have committed that we will make sure that OpenSCAP is buildable on Windows for the entire lifetime of version 1.3.x (https://github.com/OpenSCAP/openscap/blob/maint-1.3/docs/windows.md). I think it'd be possible to replace the `bzero()` by `calloc()`.

**Pull request merged**